### PR TITLE
implement new approach of caching for total coins stats endpoint

### PR DESCRIPTION
**Pull request by ice-cronus**
Endpoint fetches data only from redis.
Each child interval we're updating value in redis for last parent interval
Build initial cache for 90 days on service startup

### CODE REVIEW

File: tokenomics/balance.go [hunk 1]
```diff
@@ -352,32 +340,170 @@ func ApplyPreStaking(amount, preStakingAllocation, preStakingBonus float64) (flo
...
+		timeSeries = append(timeSeries, &TotalCoinsTimeSeriesDataPoint{Date: date})
 	}
-	if len(cachedData) == 0 {
-		return nil, nil
+	return dates, timeSeries
+}
+
+func (r *repository) cacheTotalCoins(ctx context.Context, coins []*dwh.TotalCoins) error {
+	now := time.Now()
+	if responses, err := r.db.TxPipelined(ctx, func(pipeliner redis.Pipeliner) error {
```
Comment (ice-ares) [inline tokenomics/balance.go]: Dont complicate things, just storage.Set is enough

File: tokenomics/balance.go [hunk 2]
```diff
@@ -352,32 +340,170 @@ func ApplyPreStaking(amount, preStakingAllocation, preStakingBonus float64) (flo
...
-		return errors.Wrapf(err, "failed to serialize cache value %v", summary)
+func (r *repository) totalCoinsCacheKey(date stdlibtime.Time) string {
+	return fmt.Sprintf("totalCoinStats:%v", date.Truncate(r.cfg.GlobalAggregationInterval.Parent).Format(stdlibtime.RFC3339))
+}
+
+func (r *repository) keepTotalCoinsCacheUpdated(ctx context.Context) {
+	for ctx.Err() == nil {
+		now := time.Now()
+		if now.Sub(now.Truncate(r.cfg.GlobalAggregationInterval.Child)) >= stdlibtime.Duration(float64(r.cfg.GlobalAggregationInterval.Child)*0.75) {
+			dwhCtx, cancel := context.WithTimeout(ctx, stdlibtime.Duration(0.20*float64(r.cfg.GlobalAggregationInterval.Child)))
```
Comment (ice-ares) [inline tokenomics/balance.go]: Whats that strange value? Just set it to 1min

File: tokenomics/balance.go [hunk 3]
```diff
@@ -352,32 +340,170 @@ func ApplyPreStaking(amount, preStakingAllocation, preStakingBonus float64) (flo
...
+				dates = append(dates[:dateIdx], dates[dateIdx+1:]...)
+
+				break
+			}
+		}
+	}
+	for _, d := range dates {
+		if err = backoff.RetryNotify(
+			func() error {
+				return errors.Wrapf(r.buildTotalCoinCache(ctx, d), "failed to build/init total coins cache for %v", d)
```
Comment (ice-ares) [inline tokenomics/balance.go]: Keep it simple. Infinite retry until it works

File: tokenomics/balance.go [hunk 4]
```diff
@@ -24,31 +25,21 @@ import (
...
+	dates, res.TimeSeries = r.totalCoinsDates(now, days)
+	totalCoins, err := r.getCachedTotalCoins(ctx, dates)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to SelectTotalCoins for createdAts:%#v", dates)
+		return nil, errors.Wrapf(err, "failed to getCachedTotalCoins for createdAts:%#v", dates)
 	}
 	for _, child := range res.TimeSeries {
 		for _, stats := range totalCoins {
-			if stats.CreatedAt.Equal(child.Date) {
+			dayMatch := stats.CreatedAt.Truncate(r.cfg.GlobalAggregationInterval.Parent).Equal(child.Date.Truncate(r.cfg.GlobalAggregationInterval.Parent))
```
Comment (ice-ares) [inline tokenomics/balance.go]: U need to work with hours / child aggregate, not parent. Cuz a day is actually the xx:00 hour of the next.

**Review by ice-ares — CHANGES_REQUESTED** [review]

File: tokenomics/balance.go (hunk 4 above)
Comment (ice-cronus) [reply tokenomics/balance.go]: So, we gonna store children in redis, not parents?

Comment (ice-ares) [reply tokenomics/balance.go]: Yep

Comment (ice-ares) [reply tokenomics/balance.go]: The dates that are calculated from days, are actually hours

File: tokenomics/balance.go [hunk 5]
```diff
@@ -352,32 +338,128 @@ func ApplyPreStaking(amount, preStakingAllocation, preStakingBonus float64) (flo
 	return standardAmount, preStakingAmount
 }
 
-func (r *repository) getCachedTotalCoins(ctx context.Context, days uint64) (cached *TotalCoinsSummary, err error) {
-	cachedData, err := r.db.Get(ctx, coinStatsCacheKey(days)).Bytes()
-	if err != nil && !errors.Is(err, redis.Nil) {
-		return nil, errors.Wrapf(err, "failed to get cached value for coinStats for %v", days)
+func (r *repository) totalCoinsDates(now *time.Time, days uint64, dayInterval stdlibtime.Duration) ([]stdlibtime.Time, []*TotalCoinsTimeSeriesDataPoint) {
```
Comment (ice-ares) [inline tokenomics/balance.go]: do u need that dayInterval stdlibtime.Duration?

Comment (ice-cronus) [reply tokenomics/balance.go]: Yes, that logic was initially used for building dates for fetching data on endpoint (with interval between times = one day / parent), but now we're using it for building initial cache (with interval = one hour / child) as well. We can change it to a bool flag and switch to child / parent inside instead.

File: tokenomics/balance.go [hunk 6]
```diff
@@ -352,32 +338,128 @@ func ApplyPreStaking(amount, preStakingAllocation, preStakingBonus float64) (flo
...
+		}
+	}
+	datesSameDay := make([]stdlibtime.Time, 0, childIntervalsInParent)
+	for i := range dates {
+		if i > 0 && dates[i-1].Truncate(r.cfg.GlobalAggregationInterval.Parent).Equal(dates[i].Truncate(r.cfg.GlobalAggregationInterval.Parent)) {
+			datesSameDay = append(datesSameDay, dates[i-1])
+		} else if i > 0 && len(datesSameDay) > 0 {
+			datesSameDay = append(datesSameDay, dates[i-1])
+			log.Info(fmt.Sprintf("Building total coins cache %v...", datesSameDay[0]))
+			err = errors.Wrapf(r.buildTotalCoinCache(ctx, datesSameDay...), "failed to build/init total coins cache for %v", datesSameDay)
```
Comment (ice-ares) [inline tokenomics/balance.go]: why do u call this twice?

File: tokenomics/balance.go [hunk 7]
```diff
@@ -352,32 +338,128 @@ func ApplyPreStaking(amount, preStakingAllocation, preStakingBonus float64) (flo
...
+	}
+	for _, d := range dates {
+		found := false
+		for _, stats := range totalCoins {
+			if stats.CreatedAt.Equal(d) {
+				found = true
+				break
+			}
+		}
+		if !found {
```
Comment (ice-ares) [inline tokenomics/balance.go]: why?

File: tokenomics/balance.go (hunk 6 above)
Comment (ice-cronus) [reply tokenomics/balance.go]: Infinite retry until it works. Is context approach with break better?
https://github.com/ice-blockchain/freezer/pull/79#discussion_r1435602231

Comment (ice-ares) [reply tokenomics/balance.go]: u have 2 calls to the same method, one after the other

File: tokenomics/balance.go (hunk 7 above)
Comment (ice-cronus) [reply tokenomics/balance.go]: actually not needed, cuz we're fetching on endpoint only matched data, removed

File: tokenomics/balance_total_coins.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+		return nil, errors.Wrapf(err, "failed to getCachedTotalCoins for createdAts:%#v", dates)
+	}
+	for i, child := range res.TimeSeries {
+		for _, stats := range totalCoins {
+			dayMatch := stats.CreatedAt.Equal(child.Date)
+			if dayMatch {
+				child.Standard = stats.BalanceTotalStandard
+				child.PreStaking = stats.BalanceTotalPreStaking
+				child.Blockchain = stats.BalanceTotalEthereum
+				child.Total = child.Standard + child.PreStaking + child.Blockchain
```
Comment (ice-ares) [inline tokenomics/balance_total_coins.go]: remove + child.Blockchain

File: tokenomics/balance_total_coins.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+				break
+			}
+		}
+		if i == 0 {
+			child.Date = child.Date.In(location).Add(-1 * stdlibtime.Nanosecond)
```
Comment (ice-ares) [inline tokenomics/balance_total_coins.go]: No, return it in utc

File: tokenomics/balance_total_coins.go [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+		dates              = make([]stdlibtime.Time, 0, days)
+		timeSeries         = make([]*TotalCoinsTimeSeriesDataPoint, 0, days)
+		dayInterval        = r.cfg.GlobalAggregationInterval.Parent
+		start              = now.Add(-1 * truncationInterval).Truncate(truncationInterval)
+	)
+	midnightHour := 0
+	if r.cfg.GlobalAggregationInterval.Child < stdlibtime.Hour {
+		midnightHour = now.Hour()
+	}
+	timeSinceMidnight := now.UTC().Sub(stdlibtime.Date(now.Year(), now.Month(), now.Day(), midnightHour, 0, 0, 0, stdlibtime.UTC))
```
Comment (ice-ares) [inline tokenomics/balance_total_coins.go]: What is this overengineeribg? Just Truncate to parent

File: tokenomics/balance_total_coins.go [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+	dates = append(dates, start)
+	timeSeries = append(timeSeries, &TotalCoinsTimeSeriesDataPoint{Date: start})
+	for day := uint64(0); day < days-1; day++ {
+		date := now.Add(dayInterval * -1 * stdlibtime.Duration(day)).Add(-timeSinceMidnight).Truncate(truncationInterval)
```
Comment (ice-ares) [inline tokenomics/balance_total_coins.go]: If day == 0 
   Truncate to parent
Rlse 
Truncate to child

File: tokenomics/balance_total_coins.go [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+	}
+	return dates, timeSeries
+}
+
+func (r *repository) cacheTotalCoins(ctx context.Context, coins []*dwh.TotalCoins) error {
+	val := make([]interface{ Key() string }, 0, len(coins))
+	for _, v := range coins {
+		val = append(val, v)
+	}
+	return errors.Wrapf(storage.Set(ctx, r.db, val...), "failed to set cache value for total coins: %+v", coins)
```
Comment (ice-ares) [inline tokenomics/balance_total_coins.go]: %#v

Just get a proper IDE that can highlight issues properly

File: tokenomics/balance_total_coins.go [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+	for _, cached := range alreadyCached {
+		for dateIdx, date := range dates {
+			if cached.CreatedAt.Equal(date) {
+				dates = append(dates[:dateIdx], dates[dateIdx+1:]...)
+
+				break
+			}
+		}
+	}
+	workerPool := pond.New(routinesCountToInitCoinsCacheOnStartup, 0, pond.MinWorkers(routinesCountToInitCoinsCacheOnStartup))
```
Comment (ice-ares) [inline tokenomics/balance_total_coins.go:136]: 10 goroutines in total, not min

**Review by ice-ares — CHANGES_REQUESTED** [review]

File: tokenomics/balance_total_coins.go (hunk 13 above)
Comment (ice-cronus) [reply tokenomics/balance_total_coins.go:136]: It is min and max (first arg), so 10 in total

File: tokenomics/balance_total_coins.go (hunk 11 above)
Comment (ice-cronus) [reply tokenomics/balance_total_coins.go]: not needed, first element is added before (start). But we have to duplicate last entry with parent and child, like
```
2023-12-27T17:00:00 (-1ns = 27/12)
2023-12:27T00:00:00 (-1ns = 26/12) 
```

File: tokenomics/balance_total_coins.go [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+		return nil, errors.Wrapf(err, "failed to getCachedTotalCoins for createdAts:%#v", dates)
+	}
+	for _, child := range res.TimeSeries {
+		for _, stats := range totalCoins {
+			dayMatch := stats.CreatedAt.Equal(child.Date)
+			if dayMatch {
+				child.Standard = stats.BalanceTotalStandard
+				child.PreStaking = stats.BalanceTotalPreStaking
+				child.Blockchain = stats.BalanceTotalEthereum
+				child.Total = child.Standard + child.PreStaking
```
Comment (ice-ares) [inline tokenomics/balance_total_coins.go]: Lets not do any math here. Keep em as is. Imma add the math to the dwh storage layer later on

File: tokenomics/balance_total_coins.go [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+		}
+	}
+	workerPool := pond.New(routinesCountToInitCoinsCacheOnStartup, 0, pond.MinWorkers(routinesCountToInitCoinsCacheOnStartup))
+	for _, date := range dates {
+		fetchDate := date
+		workerPool.Submit(func() {
+			err = errors.New("first try")
+			for err != nil {
+				log.Info(fmt.Sprintf("Building total coins cache for %v", fetchDate))
+				err = errors.Wrapf(r.buildTotalCoinCache(ctx, date), "failed to build/init total coins cache for %v", fetchDate)
```
Comment (ice-ares) [inline tokenomics/balance_total_coins.go]: Ctx, fetchDate

**Pull request merged**